kep/tbd-gpu-aware-scheduling/README.md

```diff
@@ -0,0 +1,408 @@
+- The Filter plugin should not affect any other scheduling decision made by the kubernetes scheduler.
+
+### Non-Goals
+- The added Filter plugin is not responsible for any scheduling decision made to schedule GPU pods on GPU nodes
+and non-GPU pods on non-GPU nodes.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+To prevent non-GPU pods from getting scheduled on GPU nodes, we plan to add a new filter 
+and prefilter plugins at Filter and PreFilter extension points respectively.
+
+![PEAKS workflow](./figs/filter_plugin_flow.png)
+
+#### PreFilter State
+Every plugin in kubernetes scheduler keeps a CycleState to store and retrieve arbitrary data. 
+CycleState uses a sync.Map to back the storage. It's aimed to optimize for the write once and 
+read many times scenarios. It is the recommended pattern used in all in-tree plugins. 
+Plugin-specific state is written once in PreFilter/PreScore and afterwards read many times in Filter/Score.
+
+In our case PreFilter will check if a given pod is requesting GPUs or not by checking resource 
+limits and will update the PreFilter State accordingly. Pod requesting non-zero GPU
+in limits will be considered as GPU requesting pod. Checking pod limit because as per [kuberetes 
+documentations](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/#using-device-plugins),
+- You can specify resource limits without specifying requests because Kubernetes will use the limit as the request value by default.
+- You can specify the resource in both limits and requests but these two values must be equal.
+- You cannot specify resource requests without specifying limits.
+
+#### Filter State
+For a given pod resource, Filter plugin will check all the nodes in the cluster. It will check
+whether the node is GPU node or not by checking node’s Allocatable field. If a node has an allocatable
+GPU resource, it will consider that node as GPU node. 'Allocatable' on a Kubernetes 
+node is defined as the amount of compute resources that are available for pods. Note that node 
+resources which have been allocated to pods running on that node will not be subtracted from the 
+node allocatable resources.
```

Comment:
Assuming there's no high pod churn, and assuming the working set of pod is relatively stable (e.g. pods tend to be long-running).
Would https://github.com/kubernetes-sigs/descheduler help in this use case?
Maybe not, let's discuss in the `Alternatives` section

Reply:
Microservices use cases have low pods churn in general. But batch workload use cases have very high pod churn.

Reply:
fine. This deserves to be recorded in the `Alternatives` section, explaining why the proposed solution is better than `descheduler`, and on which scenarios.